### ISSUE TITLE
enh(skore): Add a `decision_threshold` parameter to the `confusion_matrix` display

### ISSUE THREAD

**Issue by glemaitre** [probabl-ai/skore#2112]
### What would you like to say?

This issue is related to #2111. The entries of the confusion matrix depends on the decision threshold used to threshold the output of the `model.predict_proba`.

Showing a `confusion_matrix.frame(decision_threshold=0.1)` or `confusion_matrix.plot(decision_threshold=0.1)` would be really handy.

In terms of design, we have 2 choices:

- pre-computed a bunch of confusion matrix for different threshold ahead of time and when calling `.plot` or `.frame`, we pick-up the matrix with the decision threshold the closest to the one requested
- or only store the predictions and when calling `.plot` and `.frame`, we should threshold and then compute the matrix.

Given the current limitation on the hub side regarding the compute, the pre-computing could be the best solution at the short term.

**Comment by GaetandeCast** [comment]
Hi @glemaitre, I'm looking into how to solve the issue.

First of all, concerning the choice of design here are the pros and cons that I see with each method:
- The first method pre-computes the confusion matrix for a pre-determined set of thresholds (e.g. [0.1, 0.2, ..., 0.9]) and displays the matrix computed for the closest threshold. This allows the hub to display the matrix for any of the thresholds, without any computation required. The main drawback is that it is not flexible in the choice of threshold as only a fixed set is given as option to the user. However it is a straightforward way of seeing how the matrix can change depending on the threshold.
- The second method computes the matrix on the fly when `.plot` or `.frame` is called, based on the predicted class probabilities. For a possible integration in the hub, this requires some compute on the hub or to have the predict probas available. On the lib side, this allows the user to set a specific threshold and not default to the closest pre-computed one.

Note: I'm assuming that the intended use for the hub is to visualize how the matrix changes when the threshold moves, maybe by using a cursor on the threshold value. Please correct me if I'm wrong.

If we want a simple implementation first, I can start working on the first solution.
I am unsure of when to compute the matrices though. Should it be done at the first call to `confusion_matrix.frame()` or `confusion_matrix.plot()` and put into cache ? Or should it be when the report is initialized ? Or something else ?

cc @tuscland

**Comment by glemaitre** [comment]
> Note: I'm assuming that the intended use for the hub is to visualize how the matrix changes when the threshold moves, maybe by using a cursor on the threshold value. Please correct me if I'm wrong.

Yes it would be the UX and it means that pre-defined threshold could be good enough.

> I am unsure of when to compute the matrices though. Should it be done at the first call to confusion_matrix.frame() or confusion_matrix.plot() and put into cache ? Or should it be when the report is initialized ? Or something else ?

The idea would be to launch all the computation when creating the `Display` and to not have to make computation during `.plot` or `.frame`. Right now the `ConfusionMatrixDisplay` is not following our API (https://github.com/probabl-ai/skore/issues/2107?issue=probabl-ai%7Cskore%7C1757) but actually what would happen normally in terms of API:

- the report will call the `Display` factory called: `compute_data_for_display` (e.g. https://github.com/probabl-ai/skore/blob/main/skore/src/skore/_sklearn/_plot/metrics/roc_curve.py#L854) that then will call the `Display.__init__`.
- it means that all data are stored in the `Display`
- thus calling `.plot` or `.frame` it means that we *select* a portion of the data and just show this bit and thus we don't trigger any new computation and it is therefore efficient and don't require caching.

So if it is clear maybe the right way is to solve the following issue first: https://github.com/probabl-ai/skore/issues/2107?issue=probabl-ai%7Cskore%7C1757

I think that we can quickly solve the issue together and then start to work on solution one as a first approach.

**Comment by glemaitre** [comment]
Actually, https://github.com/scikit-learn/scikit-learn/pull/30134 just got merged where there is another strategy regarding the thresholds: all thresholds are computed given the data. It means that we could go have this strategy.

Also, it means that we should call this `confusion_matrix_at_thresholds` whenever is available but for the older scikit-learn version, we need to backport this implementation. For the backport, the amount of code could be minimum because we can create it from the `_binary_clf_curve` and just add what we are missing to have the signature of the new function.

So at the end, it is still a variant of 1. because we will match the closer threshold that the user wants which is completely fine because an intermediate threshold would not exist because from the data, we could not have it.

**Comment by glemaitre** [comment]
One thing that we will encounter during the implementation is that the `decision_threshold` will only have an impact on the binary classification case at first. It is not clear to me how do we handle the multiclass since that the confusion matrix is derived from an argmax and not some direct thresholding.

**Comment by auguste-probabl** [comment]
Blocked by #1757

**Comment by GaetandeCast** [comment]
Now that the `ConfusionMatrixDisplay` API is a bit more aligned with the rest, I've looked into implementing this.

- A first note is that I'm unsure of the necessity to backport scikit-learn's `confusion_matrix_at_thresholds`. In our situation, 
```
tns, fps, fns, tps, thresholds = confusion_matrix_at_thresholds(y_true, y_pred)
```
is equivalent to 
```
fps, tps, thresholds = _binary_clf_curve(y_true, y_pred)
fns = tps[-1] - tps
tns = fps[-1] - fps
```
I don't think it is worth checking the scikit-learn version just to avoid these two lines of code. I may be unaware of some considerations that make it worth it though, and in this situation I guess we could implement a backport of `confusion_matrix_at_thresholds` in `_sklearn_compat.py`.

- On a technical implementation detail: should we store all the confusion matrices in a new attribute `self.confusion_matrices` and keep a default one in `self.confusion_matrix` for backward compat? The default one being at the threshold closest to `0.5` ? Or keep an array of matrices stored in `self.confusion_matrix` and not introduce a new attribute?

- I think it makes sense to have an `"auto"` option for `decision_threshold` that uses the threshold that maximizes balanced accuracy, for plots and frames.

I have a first implementation that can be used for a draft PR.

**Comment by glemaitre** [comment]
> A first note is that I'm unsure of the necessity to backport scikit-learn's confusion_matrix_at_thresholds

By backporting, I meant something like:

```python
if sklearn_version < version_parse("1.8"):
    def confusion_matrix_at_thresholds(...):
        from sklearn.metrics import _binary_clf_curve

        fps, tps, thresholds = _binary_clf_curve(...)
        ...
        return tns, fps, fns, tps, thresholds
else:
    from sklearn.metrics import confusion_matrix_at_thresholds
```

It should be placed in `_sklearn_compat.py`.

In this way, we use `from skore._externals._sklearn_compat import confusion_matrix_at_thresholds` and in 2 years we only need to change the import without changing any other part of the code.

At least, it centralizes the tool needed to be back compatible across the version of scikit-learn.

**Comment by glemaitre** [comment]
> On a technical implementation detail: should we store all the confusion matrices in a new attribute self.confusion_matrices and keep a default one in self.confusion_matrix for backward compat? The default one being at the threshold closest to 0.5 ? Or keep an array of matrices stored in self.confusion_matrix and not introduce a new attribute?

For the moment we don't really care about backward compatibility. So let's take profit of it. I think that we could store the information in a long format dataframe as discussed in the last point in https://github.com/probabl-ai/skore/issues/2160#issue-3628678740

Basically each row of the dataframe is a confusion matrix and the threshold will be an extra columns. It will be aligned with the storage of that we have for the other display.

> I think it makes sense to have an "auto" option for decision_threshold that uses the threshold that maximizes balanced accuracy, for plots and frames.

The balanced accuracy is actually not a super nice metric. From what I recall, you are assigning equal weight to make a mistake which is hardly the case in reality. So I'm almost sure that we should not use this heuristic. But I don't know if I have a good one. I'm wondering if we should look at the Matthews Correlation Coefficient or just acknowledge that there is no good default. I recall this paper at the time we discuss a bit the matter in scikit-learn: https://biodatamining.biomedcentral.com/articles/10.1186/s13040-021-00244-z#:~:text=by%20D,in%20two-class

However, I don't have all the details in head anymore.
